controllers-films-users

A-Glebov opened the pull request:
1. Написали UserController для получения списка все юзеров, добавления и обновления данных юзеров

2. Написали FilmController для получения списка фильмов, добавления и обновления данных фильмов
3. Написали исключение ValidationException для выбрасывания в случае невалидных данных при добавлении или обновлении
4. Добавили NotFoundException для выбрасывания в случае невалидного id
5. Добавили тесты

A-Glebov committed:
1. Написали UserController для получения списка все юзеров, добавления и обновления данных юзеров
2. Написали FilmController для получения списка фильмов, добавления и обновления данных фильмов
3. Написали исключение ValidationException для выбрасывания в случае невалидных данных при добавлении или обновлении
4. Добавили NotFoundException для выбрасывания в случае невалидного id
5. Добавили тесты

A-Glebov committed:
1. Написали UserController для получения списка все юзеров, добавления и обновления данных юзеров
2. Написали FilmController для получения списка фильмов, добавления и обновления данных фильмов
3. Написали исключение ValidationException для выбрасывания в случае невалидных данных при добавлении или обновлении
4. Добавили NotFoundException для выбрасывания в случае невалидного id
5. Добавили тесты

A-Glebov committed:
1. Написали UserController для получения списка все юзеров, добавления и обновления данных юзеров
2. Написали FilmController для получения списка фильмов, добавления и обновления данных фильмов
3. Написали исключение ValidationException для выбрасывания в случае невалидных данных при добавлении или обновлении
4. Добавили NotFoundException для выбрасывания в случае невалидного id
5. Добавили тесты

A-Glebov committed:
1. Написали UserController для получения списка все юзеров, добавления и обновления данных юзеров
2. Написали FilmController для получения списка фильмов, добавления и обновления данных фильмов
3. Написали исключение ValidationException для выбрасывания в случае невалидных данных при добавлении или обновлении
4. Добавили NotFoundException для выбрасывания в случае невалидного id
5. Добавили тесты

BerezinYaroslav commented on src/main/java/ru/yandex/practicum/filmorate/model/Film.java:
```diff
@@ -1,12 +1,20 @@
+
 @Getter
 @Setter
+@EqualsAndHashCode
```
➕ Для сущностей вместо нескольких аннотаций можно использовать одну - `@Data`, которая включает в себя:
- `@Getter`
- `@Setter`
- `@ToString`
- `@EqualsAndHashCode`
- `@RequiredArgsConstructor`

Но нужно осторожно подходить к ее использованию. Иногда лучше использовать несколько аннотаций + ручное переопределение методов

BerezinYaroslav replied:
✅

BerezinYaroslav commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -1,7 +1,130 @@
+    @PostMapping
+    public Film create(@RequestBody Film film) {
+        log.info("Добавление фильма");
+        if (film.getName() == null || film.getName().isEmpty()) {
```
➕ Всех этих ручных проверок можно избежать, использую аннотацию `@Valid` рядом с `@RequestBody` и настроенной валидацией в классе сущности. Там есть удобные аннотации: `@NotNull` к полю класса-модели для проверки на null , `@NotBlank` — для проверки на пустую строку, `@Email` — для проверки на соответствие формату электронного адреса и другие. Подробнее можешь посмотреть в документации: https://www.baeldung.com/spring-boot-bean-validation


Для ее работы нужна новая зависимость:

```
<dependency>
    <groupId>org.springframework.boot</groupId>
    <artifactId>spring-boot-starter-validation</artifactId>
</dependency>
```

BerezinYaroslav replied:
✅

BerezinYaroslav commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -1,7 +1,130 @@
+        log.info("Добавление фильма");
+        if (film.getName() == null || film.getName().isEmpty()) {
+            String errorMessage = "Название не может быть пустым";
+            log.error("Ошибка валидации добавления фильма: {}", errorMessage);
```
✅ Верно, при выбросе ошибок тоже нужно логирование

BerezinYaroslav commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -1,7 +1,130 @@
+@RequestMapping("/films")
 public class FilmController {
+
+    private final Map<Long, Film> films = new HashMap<>();
```
✅ Хранение через мапы - самый удобный подход в этом ТЗ

BerezinYaroslav commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -1,7 +1,130 @@
+    private final Map<Long, Film> films = new HashMap<>();
+
+    @GetMapping
+    public Collection<Film> findAll() {
```
➕ Лучше возвращать `List<Film>`

BerezinYaroslav replied:
✅

BerezinYaroslav commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -1,7 +1,130 @@
+        film.setId(getNextId());
+        films.put(film.getId(), film);
+        log.info("Фильм успешно добавлен");
+        return film;
```
✅ Верно, при сохранении и обновлении стоит возвращать id сохраненной сущности или саму сущность

BerezinYaroslav commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -1,7 +1,130 @@
+    }
+
+    @PutMapping
+    public Film update(@RequestBody Film updateFilm) {
```
➕ Лучше назвать параметр updatedFilm или просто film

BerezinYaroslav replied:
✅

BerezinYaroslav commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -1,7 +1,130 @@
+
+            LocalDate newDate = updateFilm.getReleaseDate();
+            if (newDate != null) {
+                if (newDate.isBefore(LocalDate.of(1895, 12, 28))) {
```
⛔️ У тебя в обоих методах (сохраниение и обновление) дублируются проверки и выброс ошибки. Можно либо сделать валидацию в классе сущности, как я писал выше, либо вынести логику проверки фильмов в отдельный класс и в начале методов создания / обновления вызывать его. Можно назвать его validateFilm

BerezinYaroslav replied:
✅

BerezinYaroslav commented on src/main/java/ru/yandex/practicum/filmorate/controller/UserController.java:
```diff
@@ -0,0 +1,178 @@
+@Slf4j
+@RestController
+@RequestMapping("/users")
+public class UserController {
```
⛔️ По этому классу замечание то же: либо вынести валидацию в класс сущности, либо сделать отдельный метод, чтобы избежать дублирования кода. Подход с валидацией в классах считается более современным и удобным

BerezinYaroslav replied:
✅

BerezinYaroslav commented on src/main/java/ru/yandex/practicum/filmorate/controller/UserController.java:
```diff
@@ -0,0 +1,178 @@
+    private final Map<Long, User> users = new HashMap<>();
+
+    @GetMapping
+    public Collection<User> findAll() {
```
➕ Здесь тоже лучше возвращать `List<User>`

BerezinYaroslav replied:
✅

BerezinYaroslav commented on src/main/java/ru/yandex/practicum/filmorate/controller/UserController.java:
```diff
@@ -0,0 +1,178 @@
+        }
+
+        if (user.getLogin() == null || user.getLogin().isBlank() || user.getLogin().contains(" ")) {
+            String errorMessage = "Логин не может содержать пробелы";
```
⛔️ Ошибка указывает на пробелы, но в проверке сразу 3 условия

BerezinYaroslav replied:
✅

BerezinYaroslav reviewed (changes requested):
Алексей, привет!

Очень хорошая работа, но есть правки. Если будут вопросы - пиши в Пачке

Удачи 🤝

A-Glebov committed:
1. Применили аннотации из spring-boot-starter-validation вместо валидации в контроллерах.
2. Вынесли валидацию релиза фильма в класс Film
3. Добавили GlobalExceptionHandler
4. Убрали лишние тесты

BerezinYaroslav commented on src/main/java/ru/yandex/practicum/filmorate/model/Film.java:
```diff
@@ -1,12 +1,40 @@
+    @Positive
+    private Long duration;
+
+    public void dateReleaseValidate() {
```
➕ Можно и так

BerezinYaroslav reviewed (approved):
Алексей, привет)

Все супер, принято! Удачи в новом спринте 🤝
